staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -158,26 +286,100 @@ func (q *delayingType) ShutDown() {
+}
+
+// AddWithOptions is a new method that add's items to the delayed queue.
+// Calls are asynchronous so that the client can return quicker and continue with useful work.
+// The DelayingOptions give the client more control over the queueing behaviour.
+func (q *delayingType) AddWithOptions(item interface{}, opts DelayingOptions) {
+	q.sendItemToWaitingLoop(item, opts, waitForActionQueue, false)
+}
+
+// ForgetWaiting removes an item from the delayed queue, effectively cancelling its future processing.
+func (q *delayingType) ForgetWaiting(item interface{}) {
+	q.sendItemToWaitingLoop(item, DelayingOptions{}, waitForActionForget, false)
+}
+
+// sendItemToWaitingLoop facilitates the different interface methods and performs the actual add, sending,
+// forgetting an item, by sending an appropriate message through to the waitingLoop.
+func (q *delayingType) sendItemToWaitingLoop(item interface{}, opts DelayingOptions, action waitForAction, synchronous bool) {
 	// don't add if we're already shutting down
 	if q.ShuttingDown() {
 		return
 	}
 
-	q.metrics.retry()
+	// Add the item into the delaying queue (either the 'active' or 'waiting' queue)
+	waitEntry := &waitFor{
+		action:  action,
+		data:    item,
+		readyAt: q.clock.Now().Add(opts.Duration),
+		options: opts,
+	}
+	if q.logger != nil {
+		if synchronous {
+			q.log(fmt.Sprintf("client: sending message to waitingLoop: %s", waitEntry))
+		} else {
+			q.log(fmt.Sprintf("client: sending ASYNC message to waitingLoop: %s", waitEntry))
+		}
+	}
 
-	// immediately add things with no delay
-	if duration <= 0 {
-		q.Add(item)
-		return
+	// For synchronous calls we will create a channel that we can wait to be released (closed).
+	var cb chan waitingLoopResponse
+	if synchronous {
+		cb = make(chan waitingLoopResponse)
+		waitEntry.returnCh = cb
 	}
 
+	// Send the entry to the waitingLoop...
 	select {
 	case <-q.stopCh:
 		// unblock if ShutDown() is called
-	case q.waitingForAddCh <- &waitFor{data: item, readyAt: q.clock.Now().Add(duration)}:
+	case q.waitingForAddCh <- waitEntry:
+		// wait for call-back if we asked for one...
+		if synchronous {
+			select {
+			case <-q.stopCh:
+			case <-cb:
+			}
+		}
+	}
+}
+
+// log outputs a message with a Unix Time Stamp with Nanoseconds to the queue logger.
+func (q *delayingType) log(msg string) {
+	if q.logger == nil {
+		return
 	}
+	q.logger.Log(format_time(q.clock.Now()) + ": " + msg)
+}
+
+// format_time - makes a time more human parseable by placing a '.' between the seconds and the fractional part.
+func format_time(t time.Time) string {
+	un := t.UnixNano()
+	secs := t.Unix()
+	frac := un - (secs * 1000000000)
+	return strconv.FormatInt(secs, 10) + "." + strconv.FormatInt(frac, 10)
+}
+
+// waitingLoopResponse describes the message that is returned from the waiting loop when a response
+// is requested/required.  It contains the different answer types that may be in the response.
+type waitingLoopResponse struct {
```

Comment:
+1, much better to have a common struct. We can replace all the fields with pointers, to distinguish the presence of an answer and to lower the memory footprint for the nextReady.